src/isomorphic/classic/element/ReactElementValidator.js

```diff
@@ -183,18 +183,46 @@ function validatePropTypes(element) {
 var ReactElementValidator = {
 
   createElement: function(type, props, children) {
-    var validType = typeof type === 'string' || typeof type === 'function' ||
-                    (type !== null && typeof type === 'object');
-    // We warn in this case but don't throw. We expect the element creation to
-    // succeed and there will likely be errors in render.
-    if (!validType) {
-      warning(
-        false,
-        'React.createElement: type should not be null, undefined, boolean, or ' +
-          'number. It should be a string (for DOM elements) or a ReactClass ' +
-          '(for composite components).%s',
-        getDeclarationErrorAddendum()
-      );
+    var validType = true;
+    var warnInvalidType = function(inputValue) {
```

Comment:
It can be expensive to create closures in such a hot path. Can you please hoist that function outside createElement()?